Ingress Documentation (#376)
canonical/k8s-snap: docs/src/snap/explanation/ingress.md

# Ingress

In Kubernetes, understanding how inbound traffic is managed inside of your
cluster can be complex.
While there is abundant documentation, this explanation provides you with the essentials
to successfully manage your Canonical Kubernetes cluster.

## Kubernetes Pods and Services

In Kubernetes, the smallest unit is a pod, which encapsulates application containers.
Since pods are ephemeral and their IP addresses change when destroyed and restarted,
they are exposed through services.
Services offer a stable network interface by providing discoverable names and
load balancing functionality for managing a set of pods.
For further details on Kubernetes Services,

> refer to the upstream documentation: [Kubernetes Service][Service].

Reviewer: ```suggestion
refer to the [upstream Kubernetes Service documentation][Service].
```

## Ingress

> [Ingress][Ingress K8s] is a Kubernetes Resource that manages

Reviewer: is that 'R' really an 'r'?

external access by handling both HTTP and HTTPS traffic to services within your cluster.
Traffic routed through the Ingress is directed to a service,
which in turn forwards it to the relevant pod
running the desired application within a container.

The Ingress Resource lets you define rules on how traffic should get handled.

> Learn more about these rules and how to implement them here:

Reviewer: probably better to say "Refer to the [Kubernetes documentation on Ingress rules] for up to date information on the available rules and their implementation." or something like that.

[Ingress Rules][Ingress Rules].

While the Ingress Resource manages the routing rules for the incoming traffic,
the [Ingress Controller][Ingress Controller] is responsible for implementing
those rules by configuring the underlying networking infrastructure of the cluster.
Ingress does not work without an Ingress Controller.

The Ingress Controller also serves as a layer 7 (HTTP/HTTPS) load balancer
that routes traffic from outside of your cluster to services inside of your cluster.

> Please do not confuse this with the Kubernetes Service LoadBalancer type.

Reviewer 2: Maybe add a brief sentence about the difference between both.

```{kroki} ../../assets/ingress.puml
```

With Canonical Kubernetes, enabling default Ingress is easy:

> See [default Ingress][Ingress].

Reviewer: ```suggestion
See the [default Ingress guide][Ingress].
```

> At the moment of writing, enabling default Ingress will create a

Reviewer 2: This sounds a bit like "at the moment we use Cilium but that might change at any moment". Maybe just remove the specific type altogether.

[Cilium Ingress Controller][Cilium Ingress Controller] for you.
If your cluster requires different Ingress Controllers,
the responsibility of implementation falls upon you.

> 

Reviewer: I think it would be useful to add a note here which more specifically calls out that the underlying mechanism  provided by default will always operate in the same way but may not always be Cilium (if indeed, that is accurate!)

Author: So this one is a bit of a slippery slope. If we talk about it it may sound like we may change it some time. Have a look at my change and see what you think. @bschimke95 @evilnick you might have a different opinion?

Reviewer: "We reserve the right to..." sounds more like we are definitely planning to change it :joy: 

perhaps we should ask @perk

Reviewer 3: I propose the following wording here:
> The underlying mechanism provided by default is currently Cilium. However, it should always be operated through the provided CLI rather than directly. This way, we can provide the best experience for future cluster maintenance and upgrades.

See my suggestion here: https://github.com/canonical/k8s-snap/pull/376/files/76cde6aa0ac7e6d57ead5a8938e0e0fda0deb537..c585437d88e2da9c006e215369d333bb6cf9395f#r1579796674

Author: Thank you!

You will need to create the Ingress Resource,
outlining rules that direct traffic to your application's Kubernetes service.

> ## Network

Reviewer 2: Maybe put this at the beginning. This reads like a prerequisite.

Author: Sure thing!

> In order to use Canonical Kubernetes default Ingress ensure that the

Reviewer: I wonder if there is any way we can get around repeating Canonical Kubernetes quite so often? And the word 'default'. 
Maybe reword the whole paragraph

When you install Canonical Kubernetes, the default Network is automatically enabled. This is also a requirement for the default Ingress to function.

Canonical Kubernetes default network is enabled on your cluster.
This is the case by default.

Since upstream Kubernetes comes without a network provider,
it requires the use of a [network plugin][network plugin].
This network plugin facilitates communication between pods,
services, and external resources, ensuring smooth traffic flow within the cluster.
The current implementation of Canonical Kubernetes leverages a widely adopted
CNI (Container Network Interface) called [Cilium][Cilium].
If you wish to use a different network plugin
the implementation and configuration falls under your responsibility.

> Learn how to use the Canonical Kubernetes default network: See [Network][Network].

Reviewer 2: Integrate the link into the sentence as Nick suggested in another place.
```
Learn how to use the Canonical Kubernetes default network in the [networking HowTo][Network]
```

> Learn how to use the Canonical Kubernetes default network: See [Network][Network].

Reviewer: Refer to the [default Network guide] for more details

# Ingress

In Kubernetes, understanding how inbound traffic is managed inside of your
cluster can be complex.

> While there is abundant documentation, this explanation provides you with the essentials

Reviewer 4: Can you mention where? Or rephrase maybe

For further details on Kubernetes Services,
refer to the [upstream Kubernetes Service documentation][Service].

## Ingress

[Ingress][Ingress K8s] is a Kubernetes resource that manages
external access by handling both HTTP and HTTPS traffic to services within your cluster.
Traffic routed through the Ingress is directed to a service,
which in turn forwards it to the relevant pod
running the desired application within a container.

The Ingress resource lets you define rules on how traffic should get handled.
Refer to the [Kubernetes documentation on Ingress rules][Ingress Rules]
for up to date information on the available rules and their implementation.

While the Ingress resource manages the routing rules for the incoming traffic,
the [Ingress Controller][Ingress Controller] is responsible for implementing
those rules by configuring the underlying networking infrastructure of the cluster.
Ingress does not work without an Ingress Controller.

The Ingress Controller also serves as a layer 7 (HTTP/HTTPS) load balancer
that routes traffic from outside of your cluster to services inside of your cluster.
Please do not confuse this with the Kubernetes Service LoadBalancer type

> which operates at layer 4 and routes traffic directly to individual pods.

Reviewer 4: Nice

which in turn forwards it to the relevant pod
running the desired application within a container.

The Ingress resource lets you define rules on how traffic should get handled.
Refer to the [Kubernetes documentation on Ingress rules][Ingress Rules]
for up to date information on the available rules and their implementation.

While the Ingress resource manages the routing rules for the incoming traffic,
the [Ingress Controller][Ingress Controller] is responsible for implementing
those rules by configuring the underlying networking infrastructure of the cluster.
Ingress does not work without an Ingress Controller.

The Ingress Controller also serves as a layer 7 (HTTP/HTTPS) load balancer
that routes traffic from outside of your cluster to services inside of your cluster.
Please do not confuse this with the Kubernetes Service LoadBalancer type
which operates at layer 4 and routes traffic directly to individual pods.

```{kroki} ../../assets/ingress.puml
```

With Canonical Kubernetes, enabling default Ingress is easy:
See the [default Ingress guide][Ingress].
Enabling default Ingress will create an
[Ingress Controller][Cilium Ingress Controller] for you.

> The underlying mechanism provided by default will always operate in the same way,
> and will be Cilium for the foreseeable future.
> However, we do reserve the right to change it in the future if necessary.

Reviewer 3: ```suggestion
The underlying mechanism provided by default is currently Cilium. However, it should always be operated through the provided CLI rather than directly. This way, we can provide the best experience for future cluster maintenance and upgrades.
```

Author: Much better put, thanks! @perk
